src/main/java/org/apache/maven/plugins/shade/filter/MinijarFilter.java

```diff
@@ -138,13 +140,19 @@ private void removeServices( final MavenProject project, final Clazzpath cp )
                         }
 
                         try ( final BufferedReader bufferedReader =
-                           new BufferedReader( new InputStreamReader( jar.getInputStream( jarEntry ) ) ) )
+                           new BufferedReader( new InputStreamReader( jar.getInputStream( jarEntry ), UTF_8 ) ) )
                         {
                             for ( String className = bufferedReader.readLine(); className != null;
                                 className = bufferedReader.readLine() )
                             {
-                               log.info( className + " was not removed because it is a service" );
-                               removeClass( cp , className );
+                                className = className.split( "#", 2 )[0].trim();
```

Comment:
I don't understand this. When I remove this, all still works. Please explain.

Reply:
The description of `ServiceClass` says that a line in a service descriptor file may optionally be appended by a hash-separated comment. That line cuts away such optional comments.

Reply:
Okay, from [ServiceLoader](https://docs.oracle.com/javase/6/docs/api/java/util/ServiceLoader.html)

> This file contains the single line:
> 
>  com.example.impl.StandardCodecs    # Standard codecs

So let's add a test for that too, just ensure nobody removes this piece of code in the future.

Reply:
I will do that, but it will need a while -- have to find a small service JAR on Maven Central that actually contains such a line.

Reply:
I have a better proposal: let me improve the ITs with the mock repository manager, so we can create such jar ourselves.
Other option: turn the IT into a Maven multimodule project.

Reply:
fine with both, which one do you prefer?

Reply:
let's go for the latter, I'm still working on other projects and introducing MRM might take some time.

Reply:
ok I turn the IT into multi-module with self-made service JAR later this week :-)

Reply:
Done. :-)

What I do not understand is why I need https://github.com/apache/maven-shade-plugin/pull/17/commits/e4027ae805a903c74a24a9e3baca5069fd42684f. Without those explicit exclusions, maven-rat-plugin asks for a licence header in maven-compiler-plugin's `*.lst` logs!?

Reply:
looks like the IT is executed in the source-tree; instead it should be moved to target first. But I will take care of that.

Reply:
Great, thanks! :-)